### PR TITLE
Eamon/ssl target model

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

File: finetune/nn/target_blocks.py [hunk 1]
```diff
@@ -517,6 +517,185 @@ def seq_lab_internal(hidden):
...
+                    loss = tf.compat.v1.losses.sparse_softmax_cross_entropy(
+                        targets,
+                        logits,
+                        weights=weights
+                    )
+
+                    e = 0.02
+                    k = 1
+                    probs = tf.stop_gradient(tf.nn.softmax(all_logits))
+                    adv_vector = tf.random.uniform(shape=tf.shape(preturb_target),
```
Comment (benleetownsend) [inline finetune/nn/target_blocks.py]: Check that this vector is of a reasonable variance compared to the features.

File: finetune/nn/target_blocks.py [hunk 2]
```diff
@@ -517,6 +517,185 @@ def seq_lab_internal(hidden):
...
+                        #                                 [tf.shape(hidden),
+                        #                                  tf.shape(preturbed_hidden),
+                        #                                  tf.shape(adv_logits),
+                        #                                  tf.shape(adv_probs),
+                        #                                  tf.shape(adv_loss),
+                        #                                  tf.shape(gradient),
+                        #                                  tf.shape(adv_vector)],
+                        #                                summarize=100)
+
+                    preturbed_hidden = preturb_target + adv_vector
```
Comment (benleetownsend) [inline finetune/nn/target_blocks.py]: check the variances of this these vectors to check that adv_vector is noise and not the dominant signal.

File: finetune/nn/target_blocks.py [hunk 3]
```diff
@@ -517,6 +517,185 @@ def seq_lab_internal(hidden):
...
+                    adv_loss = kl_div(probs, adv_probs, sample_weight=mask)
+                    # adv_loss = tf.compat.v1.Print(adv_loss,
+                    #                                 [tf.shape(preturbed_hidden),
+                    #                                  tf.shape(adv_logits),
+                    #                                  tf.shape(adv_probs),
+                    #                                  tf.shape(adv_loss),
+                    #                                  tf.shape(probs)],
+                    #                                 summarize=100)
+                    adv_loss = tf.reduce_sum(adv_loss) / batch_size
+                    loss += 0.6 * adv_loss
```
Comment (benleetownsend) [inline finetune/nn/target_blocks.py]: Check magnitudes of the losses to verify that adv_loss is not dominating.

File: finetune/nn/target_blocks.py [hunk 4]
```diff
@@ -517,6 +517,185 @@ def seq_lab_internal(hidden):
...
+                    probs = tf.stop_gradient(tf.nn.softmax(all_logits))
+                    adv_vector = tf.random.uniform(shape=tf.shape(preturb_target),
+                                                   dtype=tf.float32)
+                    adv_vector = e * tf.nn.l2_normalize(adv_vector, axis=-1)
+                    batch_size = tf.cast(tf.shape(preturb_target)[0], tf.float32)
+                    kl_div = tf.losses.KLDivergence(reduction=tf.keras.losses.Reduction.NONE)
+                    mask = tf.sequence_mask(all_lengths,
+                                            maxlen=tf.math.reduce_max(all_lengths),
+                                            dtype=tf.float32)
+                    for _ in range(k):
```
Comment (benleetownsend) [inline finetune/nn/target_blocks.py]: I believe starting with k = 0 should give you something functional albeit without an optimal adv_vector

File: finetune/nn/target_blocks.py [hunk 5]
```diff
@@ -517,6 +517,185 @@ def seq_lab_internal(hidden):
...
+                        #                                  tf.shape(adv_probs),
+                        #                                  tf.shape(adv_loss),
+                        #                                  tf.shape(gradient),
+                        #                                  tf.shape(adv_vector)],
+                        #                                summarize=100)
+
+                    preturbed_hidden = preturb_target + adv_vector
+                    adv_logits = out_fn(preturbed_hidden)
+                    adv_probs = tf.nn.softmax(adv_logits)
+                    adv_loss = kl_div(probs, adv_probs, sample_weight=mask)
```
Comment (benleetownsend) [inline finetune/nn/target_blocks.py]: Worth checking that adv_loss is going up each step when k > 1

File: finetune/base_models/bert/modeling.py [hunk 6]
```diff
@@ -181,36 +182,40 @@ def __init__(
...
-                    initializer_range=config.initializer_range,
-                    max_position_embeddings=config.max_position_embeddings,
-                    dropout_prob=config.hidden_dropout_prob,
-                    roberta=roberta,
-                    pos_injection=config.pos_injection,
-                    positional_channels=config.positional_channels,
-                    reading_order_decay_rate=reading_order_decay_rate,
-                    anneal_reading_order=config.anneal_reading_order,
-                )
+            if not (embeddings is None):
```
Comment (lhz1029) [inline finetune/base_models/bert/modeling.py]: `if embeddings is not None`

File: finetune/base_models/bert/modeling.py [hunk 7]
```diff
@@ -287,6 +292,9 @@ def get_embedding_output(self):
     def get_embedding_table(self):
         return self.embedding_table
 
+    def get_embedding_output(self):
```
Comment (lhz1029) [inline finetune/base_models/bert/modeling.py]: I believe this method already exists a few lines up

File: finetune/input_pipeline.py [hunk 8]
```diff
@@ -173,7 +173,7 @@ def chunked_and_tokenized_dataset():
         else:
             tqdm_mode = "train"
 
-        if input_mode == InputMode.PREDICT or not has_targets(generator_fn):
+        if input_mode == InputMode.PREDICT or not has_targets(chunked_and_tokenized_dataset):
```
Comment (lhz1029) [inline finetune/input_pipeline.py:177]: thanks for the fix!

File: finetune/nn/target_blocks.py [hunk 9]
```diff
@@ -517,6 +519,299 @@ def seq_lab_internal(hidden):
...
+                        adv_vector = e * tf.nn.l2_normalize(gradient, axis=-1)
+                        adv_vector = tf.reshape(adv_vector,
+                                                tf.shape(adv_vector)[1:])
+                        adv_vector = tf.stop_gradient(adv_vector)
+
+                    adv_logits = out_fn(adv_vector)
+                    adv_probs = tf.nn.softmax(adv_logits)
+                    adv_loss = kl_div(probs, adv_probs, sample_weight=mask)
+                    adv_loss = tf.reduce_sum(adv_loss) / batch_size
+                    loss += 1 * adv_loss
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py]: I'm assuming the `1 *` part is to play around with weighting for adv_loss?

**Review by lhz1029 — COMMENTED** [review]
Nothing stands out as problematic. Like mentioned earlier, logging and isolating diffs seems like the most promising next steps--good luck!

File: finetune/target_models/semi_suprevised.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+    def pad_and_concat_batches(self, a, b, targets):
+        def pad_batch(batch, final_size):
+            pad_size = final_size - tf.shape(batch)[1]
+            padding = [[0, 0], [0, pad_size]]
+            return tf.pad(batch, padding)
+        a_len, b_len = tf.shape(a)[1], tf.shape(b)[1]
+        a_less = lambda: (pad_batch(a, b_len), b)
+        b_less = lambda: (a, pad_batch(b, a_len))
+        final_len = tf.cond(tf.less(a_len, b_len),
```
Comment (benleetownsend) [inline finetune/target_models/semi_suprevised.py:154]: tf.maximum??

**Comment by benleetownsend** [comment]
Are we going to merge this?

**Pull request closed without merging**